velox/type/Type.cpp

```diff
@@ -443,15 +443,27 @@ bool RowType::equivalent(const Type& other) const {
 }
 
 bool RowType::equals(const Type& other) const {
-  if (!this->equivalent(other)) {
+  auto& otherTyped = other.asRow();
+  if ((&other != this) &&
+      (!Type::hasSameTypeId(other) || otherTyped.size() != size())) {
+    // Nonequivalent
     return false;
   }
-  auto& otherTyped = other.asRow();
+
   for (size_t i = 0; i < size(); ++i) {
     // todo: case sensitivity
     if (nameOf(i) != otherTyped.nameOf(i)) {
       return false;
     }
+    if (childAt(i)->kind() == TypeKind::ROW) {
```

Comment:
This 'if' is not needed. Just call child->equals unconditionally.

Reply:
I see, you can then use operator==.

Reply:
Only OpaqueType and RowType have operator== reloaded, for other types it simply compares type address which wouldn't work. Shall we define the default operator== for Type as `equivalent` maybe?